### PR TITLE
Support data citation improvement

### PR DESCRIPTION
**Pull request by LeonardoGonzales**

### CODE REVIEW

File: spark-indexer/src/main/java/org/uniprot/store/spark/indexer/literature/LiteratureDocumentsToHDFSWriter.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        this.parameter = parameter;
+    }
+
+    @Override
+    public void writeIndexDocumentsToHDFS() {
+        LiteratureRDDTupleReader literatureReader = new LiteratureRDDTupleReader(parameter);
+
+        // load literature with abstract JavaPairRDD<citationId, Literature>
+        JavaPairRDD<String, Literature> literature = literatureReader.load();
+
```
Comment (ahmadshadab) [inline spark-indexer/src/main/java/org/uniprot/store/spark/indexer/literature/LiteratureDocumentsToHDFSWriter.java:41]: To be consistent with other variable name, we should call literature as literatureRDD

File: spark-indexer/src/main/java/org/uniprot/store/spark/indexer/literature/LiteratureDocumentsToHDFSWriter.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                        .fullOuterJoin(literature)
+                        .mapValues(new LiteratureUniProtKBJoin())
+                        .leftOuterJoin(communityStatsRDD)
+                        .mapValues(new LiteratureEntryStatisticsJoin(StatisticsType.COMMUNITY))
+                        .leftOuterJoin(computationalStatsRDD)
+                        .mapValues(
+                                new LiteratureEntryStatisticsJoin(StatisticsType.COMPUTATIONALLY))
+                        .mapValues(new LiteratureEntryToDocumentMapper())
+                        .values();
+
```
Comment (ahmadshadab) [inline spark-indexer/src/main/java/org/uniprot/store/spark/indexer/literature/LiteratureDocumentsToHDFSWriter.java:69]: wow, lot of joins. :-) I am thinking how are we going to achieve this in SpringBatch.

File: spark-indexer/src/main/java/org/uniprot/store/spark/indexer/main/verifiers/VerifyCitationData.java [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+            String releaseInputDir = getInputReleaseDirPath(config, jobParameter.getReleaseName());
+            String literaturePath = releaseInputDir + "literature/" + file;
+            sparkContext.hadoopConfiguration().set("textinputformat.record.delimiter", SPLITTER);
+            long count =
+                    sparkContext
+                            .textFile(literaturePath, 1000)
+                            .mapToPair(new LiteratureFileMapper())
+                            .count();
+            log.info("Number of entries: {}", count);
+            /*            List<Tuple2<String, Literature>> literature = sparkContext.textFile(literaturePath, 1000)
```
Comment (ahmadshadab) [inline spark-indexer/src/main/java/org/uniprot/store/spark/indexer/main/verifiers/VerifyCitationData.java:51]: If we don't need the below commented code please delete it

File: spark-indexer/src/main/java/org/uniprot/store/spark/indexer/main/verifiers/mapper/UniProtKBPublicationMapper.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+import org.uniprot.core.flatfile.writer.LineType;
+import org.uniprot.core.uniprotkb.UniProtKBReference;
+
+/**
+ * Maps an entry string to an iterator of tuples with values <accession, MappedReference>.
+ *
+ * <p>Created 18/01/2021
+ *
+ * @author Edd
+ */
```
Comment (ahmadshadab) [inline spark-indexer/src/main/java/org/uniprot/store/spark/indexer/main/verifiers/mapper/UniProtKBPublicationMapper.java:31]: I think, we should delete this file/

**Review by ahmadshadab — COMMENTED** [review]
Please note that i didn't see SpringBatch related change to index Literature. I think, it is intentional.

File: spark-indexer/src/main/java/org/uniprot/store/spark/indexer/main/verifiers/VerifyCitationData.java (hunk 3 above)
Comment (LeonardoGonzales) [reply spark-indexer/src/main/java/org/uniprot/store/spark/indexer/main/verifiers/VerifyCitationData.java:51]: This was a test class. I deleted it

File: spark-indexer/src/main/java/org/uniprot/store/spark/indexer/main/verifiers/mapper/UniProtKBPublicationMapper.java (hunk 4 above)
Comment (LeonardoGonzales) [reply spark-indexer/src/main/java/org/uniprot/store/spark/indexer/main/verifiers/mapper/UniProtKBPublicationMapper.java:31]: This was a test class. I deleted it

File: spark-indexer/src/main/java/org/uniprot/store/spark/indexer/common/converter/UniProtKBReferencesConverter.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+import org.uniprot.core.flatfile.parser.impl.rx.RxLineObject;
+import org.uniprot.core.flatfile.writer.LineType;
+import org.uniprot.core.uniprotkb.UniProtKBReference;
+
+/**
+ * @author lgonzales
+ * @since 24/03/2021
+ */
+public class UniProtKBReferencesConverter implements Serializable {
+
```
[the remark this answers is not in the document]
Comment (LeonardoGonzales) [reply spark-indexer/src/main/java/org/uniprot/store/spark/indexer/common/converter/UniProtKBReferencesConverter.java:29]: Test class. I deleted it

File: common-job/src/main/java/org/uniprot/store/converter/LiteratureDocumentConverter.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        byte[] literatureByte = getLiteratureObjectBinary(entry);
+        builder.literatureObj(literatureByte);
+
+        log.debug("LiteratureLoadProcessor entry: " + entry);
+        return builder.build();
+    }
+
+    private List<String> getCitationsWith(LiteratureStatistics statistics) {
+        List<String> citationsWith = new ArrayList<>();
+        if (statistics.hasReviewedProteinCount()) {
```
[the remark this answers is not in the document]
Comment (LeonardoGonzales) [reply common-job/src/main/java/org/uniprot/store/converter/LiteratureDocumentConverter.java:82]: sonar will complain. used just in one place + tests + config file

**Pull request merged**